pulp_smash/tests/pulp3/file/api_v3/test_download_content.py

```diff
@@ -0,0 +1,103 @@
+        content served by pulp can be downloaded.A distribution serves a
+        publication which was created from a repository version by a
+        publisher.Furthermore a distribution defines at which relative path on
+        Pulp's web server the publication will be served.The ``base_path`` of a
+        distribution is appended to ``/content/`` to form the full path that
+        content will be served.
+
+        This test targets the following issue:
+
+        * `Pulp #2895 <https://pulp.plan.io/issues/2895>`_
+        * `Pulp Smash #872 <https://github.com/PulpQE/pulp-smash/issues/872>`_
+
+        Do the following:
+
+        1. Create necessary resources to server content in Pulp.
+           Repository, importer, distribution, publisher and publication.
+        2. Read the checksum of content units present in the repository.
+        3. Randomly select a unit to be downloaded.
+        4. Assert that the checksum of the downloaded unit is equal to
+           checksum of the unit in repository.
+        """
+        cfg = config.get_config()
+        client = api.Client(cfg, api.json_handler)
+        client.request_kwargs['auth'] = get_auth()
+        repo = client.post(REPO_PATH, gen_repo())
+        self.addCleanup(client.delete, repo['_href'])
+        body = gen_importer()
+        body['feed_url'] = urljoin(FILE_FEED_URL, 'PULP_MANIFEST')
+        importer = client.post(FILE_IMPORTER_PATH, body)
+        self.addCleanup(client.delete, importer['_href'])
+        sync_repo(cfg, importer, repo)
+        unit_checksums = self.get_unit_checksums(
+            read_repo_content(repo)['results']
+        )
+
+        # Create a publisher.
+        publisher = client.post(FILE_PUBLISHER_PATH, gen_publisher())
+        self.addCleanup(client.delete, publisher['_href'])
+
+        # Create a publication.
+        publication = publish_repo(cfg, publisher, repo)
+        self.addCleanup(client.delete, publication['_href'])
+
+        # Create a distribution.
+        body = gen_distribution()
+        body['publication'] = publication['_href']
+        distribution = client.post(DISTRIBUTION_PATH, body)
+        self.addCleanup(client.delete, distribution['_href'])
+
+        # Download a random unit and make assertion about its checksum.
+        base_url = urljoin(CONTENT_BASE_PATH, distribution['base_path'] + '/')
+        unit_name = choice(get_content_unit_names(repo))
+        unit_url = urljoin(cfg.get_base_url(), urljoin(base_url, unit_name))
+        self.assertEqual(
+            unit_checksums[unit_name],
+            hashlib.sha256(utils.http_get(unit_url)).hexdigest()
+        )
+
+    @staticmethod
+    def get_unit_checksums(content_units):
+        """Return a dict with the name and checksum of a given units."""
+        unit_checksums = {}
+        for content_unit in content_units:
+            unit_checksums[content_unit['path']] = content_unit['digest']
+        return unit_checksums
```

Comment:
This can be a dict comprehension:

```python
return {
    content_unit['path']: content_unit['digest']
    for content_unit in content_units
}
```

Or, this method can be dropped in favor of an inline dict comprehension, above:

```python
unit_checksums = {
    content_unit['path']: content_unit['digest']
    for content_unit in read_repo_content(repo)['results']
}
```